packages/alfa-rules/src/common/predicate/is-visible.ts

```diff
@@ -33,7 +35,20 @@ export function isVisible(device: Device, context?: Context): Predicate<Node> {
       }
 
       return true;
-    }
+    },
+    // non-replaced elements with no visible child are not visible
+    // replaced elements are assumed to be replaced by something visible.
+    not(
+      and(
+        isElement,
+        and(not(isReplaced), (element) =>
+          every(
+            element.children({ nested: true, flattened: true }),
+            not(isVisible(device, context))
```

Comment:
The complexity of `isVisible()` is currently something along the lines of `O(d)`, with `d` being the depth of the node, as it walks all the way up the ancestor chain. If we evaluate it for every child of the node as well, we'll end up with something along the lines of `O(d * n)`, with `n` being the size of the subtree of the node. That seems... excessive 🙈 We should probably throw a `Cache` in there somewhere to keep the complexity at `O(d)`.

Reply:
Gaaah 🙈